Code/GraphMol/Wrap/rough_test.py

```diff
@@ -6249,28 +6237,107 @@ def test_github3553(self):
       self.assertTrue(pval in txt)
 
   def testMolzip(self):
-    tests = [
-      ["C[*:1]", "N[*:1]", "CN", Chem.MolzipParams()]
-      ]
+    tests = [["C[*:1]", "N[*:1]", "CN", Chem.MolzipParams()]]
     for a, b, res, params in tests:
-      self.assertEqual(Chem.CanonSmiles(res), Chem.MolToSmiles(
-        Chem.molzip(Chem.MolFromSmiles(a), Chem.MolFromSmiles(b), params)))
+      self.assertEqual(
+        Chem.CanonSmiles(res),
+        Chem.MolToSmiles(Chem.molzip(Chem.MolFromSmiles(a), Chem.MolFromSmiles(b), params)))
 
     # multiple arg test
     a = Chem.MolFromSmiles('C=C[1*]')
     b = Chem.MolFromSmiles('O/C=N/[1*]')
     p = Chem.MolzipParams()
     p.label = Chem.MolzipLabel.Isotope
-    c = Chem.molzip(a,b, p)
+    c = Chem.molzip(a, b, p)
     self.assertEqual(Chem.MolToSmiles(c), 'C=C/N=C/O')
 
     # single argument test
     a = Chem.MolFromSmiles('C=C[1*].O/C=N/[1*]')
     p = Chem.MolzipParams()
     p.label = Chem.MolzipLabel.Isotope
-    c = Chem.molzip(a,p)
+    c = Chem.molzip(a, p)
     self.assertEqual(Chem.MolToSmiles(c), 'C=C/N=C/O')
-    
+
+  def testContextManagers(self):
+    from rdkit import RDLogger
+    RDLogger.DisableLog('rdApp.*')
+    from io import StringIO
+    fileN = os.path.join(RDConfig.RDBaseDir, 'Code', 'GraphMol', 'Wrap', 'test_data',
+                         'github3553.sdf')
+    with Chem.SDMolSupplier(fileN) as suppl:
```

Comment:
What happens if I do:

    suppl1 = Chem.SDMolSupplier(fileN)
    with suppl1 as suppl2:
        for mol in suppl2:
            pass

? Basically, are suppl1 and suppl2 the same Python object (does `id(suppl1) == id(suppl2)`)? It looks like in the implementation, they wrap the same C++ object, but may not be the same Python object. I'd be somewhat worried about cleanup/garbage collection if this generates two Python objects which both manage the same C++ object.

I _think_ you want the `__enter__()` method to be:

    PyObject* MolIOEnter(PyObject* self) {
        Py_INCREF(self);   
        return self;
    }

... if that's possible in boost::python.

Reply:
`suppl1` and `suppl2` are certainly the same c++ object, but they aren't the same python object:
```
n [5]: suppl = Chem.SDMolSupplier('./NCI_aids_few.sdf')

In [6]: with suppl as suppl2:
   ...:     ms = [x for x in suppl2 if x is not None]
   ...: 

In [7]: suppl2 is suppl
Out[7]: False
```
As for lifetime managemnet: this should already be handled by the fact that the code is using `python::return_internal_reference` in `__enter__`

Reply:
great! I think that the other does work, though, fyi.

Also - for the exit functions - does something like this work:

    .def("__exit__", &MolIOExit<LocalForwardSDMolSupplier *>)

I ask because it would read tidier.

Reply:
nice one. I've been doing that using C-style casts forever since it works, but this is definitely both shorter and clearer.